packages/request-manager/src/torIdentities.ts

```diff
@@ -0,0 +1,83 @@
+import http from 'http';
+import https from 'https';
+import { SocksProxyAgent } from 'socks-proxy-agent';
+import { RequestInit } from 'node-fetch';
+import { TorController } from './controller';
+
+interface TorRequestOptions {
+    identity?: string;
+}
+
+export class TorIdentities {
+    static torController: TorController;
+    static identities: { [key: string]: SocksProxyAgent };
+
+    public static init(torController: TorController) {
+        this.torController = torController;
+        this.identities = {};
+        // When initializing create default identity, so it is ready to use.
+        this.getIdentity('default');
+    }
+
+    private static getIdentity(identity: string): SocksProxyAgent {
+        if (!this.identities[identity]) {
+            this.identities[identity] = new SocksProxyAgent({
+                hostname: this.torController.options.host,
+                port: this.torController.options.port,
+                userId: identity,
+                password: identity,
+            });
+        }
+        return this.identities[identity];
+    }
+
+    public static fetch(url: string, init: RequestInit, torRequestOptions?: TorRequestOptions) {
+        if (!this.torController) {
+            throw new Error('TorIdentities is not initialized');
+        }
+        const requestUrl = new URL(url);
+        const agent = this.getIdentity(torRequestOptions?.identity || 'default');
+        const body = init?.body;
+        const options = {
+            method: init?.method || 'GET',
+            hostname: requestUrl.hostname,
+            port: requestUrl.port,
+            path: requestUrl.pathname + requestUrl.search,
+            headers: (init?.headers || {}) as http.IncomingHttpHeaders,
+            agent,
+            body,
```

Comment:
I think we are going to need more options here, such as `signal`. Any objections to spreading `init` argument here?

Reply:
But actually, you we are translating window.fetch options to http.request options, am I right? So it is probably better to do it explicitly

```
interface RequestInit {
    /** A BodyInit object or null to set request's body. */
    body?: BodyInit | null;
    /** A string indicating how the request will interact with the browser's cache to set request's cache. */
    cache?: RequestCache;
    /** A string indicating whether credentials will be sent with the request always, never, or only when sent to a same-origin URL. Sets request's credentials. */
    credentials?: RequestCredentials;
    /** A Headers object, an object literal, or an array of two-item arrays to set request's headers. */
    headers?: HeadersInit;
    /** A cryptographic hash of the resource to be fetched by request. Sets request's integrity. */
    integrity?: string;
    /** A boolean to set request's keepalive. */
    keepalive?: boolean;
    /** A string to set request's method. */
    method?: string;
    /** A string to indicate whether the request will use CORS, or will be restricted to same-origin URLs. Sets request's mode. */
    mode?: RequestMode;
    /** A string indicating whether request follows redirects, results in an error upon encountering a redirect, or returns the redirect (in an opaque fashion). Sets request's redirect. */
    redirect?: RequestRedirect;
    /** A string whose value is a same-origin URL, "about:client", or the empty string, to set request's referrer. */
    referrer?: string;
    /** A referrer policy to set request's referrerPolicy. */
    referrerPolicy?: ReferrerPolicy;
    /** An AbortSignal to set request's signal. */
    signal?: AbortSignal | null;
    /** Can only be null. Used to disassociate request from any Window. */
    window?: null;
}

interface ClientRequestArgs {
        signal?: AbortSignal | undefined;
        protocol?: string | null | undefined;
        host?: string | null | undefined;
        hostname?: string | null | undefined;
        family?: number | undefined;
        port?: number | string | null | undefined;
        defaultPort?: number | string | undefined;
        localAddress?: string | undefined;
        socketPath?: string | undefined;
        /**
         * @default 8192
         */
        maxHeaderSize?: number | undefined;
        method?: string | undefined;
        path?: string | null | undefined;
        headers?: OutgoingHttpHeaders | undefined;
        auth?: string | null | undefined;
        agent?: Agent | boolean | undefined;
        _defaultAgent?: Agent | undefined;
        timeout?: number | undefined;
        setHost?: boolean | undefined;
        // https://github.com/nodejs/node/blob/master/lib/_http_client.js#L278
        createConnection?: ((options: ClientRequestArgs, oncreate: (err: Error, socket: Socket) => void) => Socket) | undefined;
        lookup?: LookupFunction | undefined;
    }
```

Reply:
buuut, node-fetch actually supports 
```    agent?: RequestOptions['agent'] | ((parsedUrl: URL) => RequestOptions['agent']);
```
so we could maybe use node-fetch instead of http.request? or is there any specific reason why we cant?

Reply:
It does not apply anymore so I am resolving this.